### PR TITLE
Two arity version of reduce

### PR DESCRIPTION
**Pull request by thomasmulvaney**
Lets you do `(reduce + (range 10))` as mentioned in #175

### CODE REVIEW

File: pixie/stdlib.pxi [hunk 1]
```diff
@@ -166,8 +166,11 @@
                           (recur)))))))))
 
 
-(def reduce (fn [rf init col]
-              (-reduce col rf init)))
+(def reduce (fn 
+              ([rf col]
+               (reduce rf (first col) col))
```
Comment (halgari) [inline pixie/stdlib.pxi]: I think this should be `(reduce rf (first col) (next col))`, Otherwise this happens: `(= (reduce + [1 2]) 4)`

**Comment by heyLu** [comment]
An alternative would be the following:

```clojure
(def reduce
  (fn
    ([rf col]
      (reduce rf (rf) col))
    ([rf init col]
      (-reduce col rf init))))
```

This would be a break from Clojure's behaviour, which [we have talked about before](https://botbot.me/freenode/pixie-lang/search/?q=rich), I'm not sure if we still want that.

Any opinions?

**Comment by thomasmulvaney** [comment]
Yeah i was thinking that could be nice too. But what about functions like
conj which don't have a meaningful identity element? The advantage of your
approach is a better error could be returned, for example "Function has no
0 arity version."

**Comment by halgari** [comment]
And that's the crux of the problem. It's not really useful, except for functions like + and *. And for those functions, why not just go with the (rf) approach? So yeah, I'm in favor of heyLu's version.

**Comment by thomasmulvaney** [comment]
Yeah, the approach given by @heyLu seems best to me too. So I will close this pull request. 

To deal with the suggested approach should we add an `-arity?` method to `IFn` so we can do something along the lines of:

```clojure
(if (arity? rf 0) 
   (rf) 
   (throw "Reducing function must support 0 args"))
```

**Comment by halgari** [comment]
It'd be better to just not have it segfault, I'll dig into it and write a PR to throw an exception on the wrong number of args.

**Pull request closed without merging**